File: cpp_merge/packages/integration/otaku/src/plugins/plugin-gamification/src/providers/pointsProvider.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/otaku/src/plugins/plugin-gamification/src/providers/pointsProvider.h"

std::shared_ptr<Provider> pointsProvider = object{
    object::pair{std:("name"), std:("USER_POINTS")}, 
    object::pair{std:("description"), std:("Provides current user points, level, and streak for agent awareness")}, 
    object::pair{std:("dynamic"), true}, 
    object::pair{std:("get"), [=](auto runtime, auto message, auto state = undefined) mutable
    {
        try
        {
            auto gamificationService = as<std::shared_ptr<GamificationService>>(runtime->getService(std:("gamification")));
            if (!gamificationService) {
                return object{
                    object::pair{std:("text"), std:("Gamification service not available")}, 
                    object::pair{std:("values"), object{}}
                };
            }
            auto summary = std::async([=]() { gamificationService->getUserSummary(message->entityId); });
            return object{
                object::pair{std:("text"), std:("User has ") + summary->allTimePoints->toLocaleString() + std:(" total points (Level: ") + summary->levelName + std:(", Streak: ") + summary->streakDays + std:(" days)")}, 
                object::pair{std:("values"), object{
                    object::pair{std:("points"), summary->allTimePoints}, 
                    object::pair{std:("weeklyPoints"), summary->weeklyPoints}, 
                    object::pair{std:("level"), summary->level}, 
                    object::pair{std:("levelName"), summary->levelName}, 
                    object::pair{std:("streak"), summary->streakDays}
                }}, 
                object::pair{std:("data"), object{
                    object::pair{std:("points"), summary->allTimePoints}, 
                    object::pair{std:("weeklyPoints"), summary->weeklyPoints}, 
                    object::pair{std:("level"), summary->level}, 
                    object::pair{std:("levelName"), summary->levelName}, 
                    object::pair{std:("streak"), summary->streakDays}
                }}
            };
        }
        catch (const any& error)
        {
            return object{
                object::pair{std:("text"), std:("Unable to fetch user points")}, 
                object::pair{std:("values"), object{}}
            };
        }
    }
    }
};

void Main(void)
{
}

MAIN

File: otaku/src/plugins/plugin-gamification/src/providers/pointsProvider.ts
import type { Provider, IAgentRuntime, Memory, State, ProviderResult } from '@elizaos/core';
import { GamificationService } from '../services/GamificationService';

export const pointsProvider: Provider = {
  name: 'USER_POINTS',
  description: 'Provides current user points, level, and streak for agent awareness',

  dynamic: true,
  get: async (runtime: IAgentRuntime, message: Memory, state?: State): Promise<ProviderResult> => {
    try {
      const gamificationService = runtime.getService('gamification') as GamificationService;
      if (!gamificationService) {
        return {
          text: 'Gamification service not available',
          values: {},
        };
      }

      const summary = await gamificationService.getUserSummary(message.entityId);

      return {
        text: `User has ${summary.allTimePoints.toLocaleString()} total points (Level: ${summary.levelName}, Streak: ${summary.streakDays} days)`,
        values: {
          points: summary.allTimePoints,
          weeklyPoints: summary.weeklyPoints,
          level: summary.level,
          levelName: summary.levelName,
          streak: summary.streakDays,
        },
        data: {
          points: summary.allTimePoints,
          weeklyPoints: summary.weeklyPoints,
          level: summary.level,
          levelName: summary.levelName,
          streak: summary.streakDays,
        },
      };
    } catch (error) {
      return {
        text: 'Unable to fetch user points',
        values: {},
      };
    }
  },
};

